executor/projection.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+	"golang.org/x/net/context"
+)
+
+// This file contains the implementation of the physical Projection Operator:
+//
+// https://en.wikipedia.org/wiki/Projection_(relational_algebra)
+//
+// For parallel implementation, a "projectionInputFetcher" fetches input Chunks
+// and dispatches them to many "projectionWorker"s, every "projectionWorker"
+// picks an input Chunk and do the projection operation then output the result
+// Chunk to the main thread: "ProjectionExec.Next()" function.
+// NOTE:
+// 1. number of "projectionWorker"s is controlled by the global session
+//    variable "tidb_projection_concurrency".
+// 2. if "tidb_projection_concurrency" is 0, we use the old non-parallel
+//    implementation.
+
+type projectionInput struct {
+	chk          *chunk.Chunk
+	targetWorker *projectionWorker
+}
+
+type projectionOutput struct {
+	chk  *chunk.Chunk
+	done chan error
+}
+
+// ProjectionExec represents a select fields executor.
+type ProjectionExec struct {
+	baseExecutor
+
+	evaluatorSuit    *expression.EvaluatorSuit // dispatched to projection workers.
+	calculateNoDelay bool
+
+	prepared   bool
+	finishCh   chan struct{}
+	outputCh   chan *projectionOutput
+	fetcher    projectionInputFetcher
+	numWorkers int64
+	workers    []*projectionWorker
+}
+
+// Open implements the Executor Open interface.
+func (e *ProjectionExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	// For now a Projection can not be executated vectorially only because it
+	// contains "SetVar" or "GetVar" functions, in this scenario this
+	// Projection can not be executed parallelly by more than 1 worker as well.
+	if e.numWorkers > 0 && !e.evaluatorSuit.Vectorizable() {
+		e.numWorkers = 1
```

Comment:
why not set e.numWorkers = 0?

Reply:
set it to 1 we can:
1. pull child row batch 2 and execute child row batch 1 concurrently
2. keep the same behavior with it set to 0